docs/hub/how-to-downstream.md

```diff
@@ -0,0 +1,53 @@
+# How to integrate downstream methods in your library
+
+In the tutorial, you learned some basic functions for integrating the Hub into your library. Functions that allow your library to download files from the Hub are referred to as *downstream* functions. This guide introduces additional downstream methods you can integrate in your library. You will learn how to:
+
+* Download a file without caching it on your disk.
+* Download a file and cache it on your disk.
+* Download all the files in a repository.
+
+## `hf_hub_url`
+
+Use `hf_hub_url` to retrieve the URL of a specific file to download by providing a `filename`.
+
+![/docs/assets/hub/repo.png](/docs/assets/hub/repo.png)
+
+```python
+>>> from huggingface_hub import hf_hub_url
+>>> hf_hub_url(repo_id="lysandre/arxiv-nlp", filename="config.json")
+'https://huggingface.co/lysandre/arxiv-nlp/resolve/main/config.json'
+```
+
+Specify a particular file version by providing the file revision. The file revision can be a branch, a tag, or a commit hash:
+
+```python
+>>> hf_hub_url(repo_id="lysandre/arxiv-nlp", filename="config.json", revision="877b84a8f93f2d619faa2a6e514a32beef88ab0a")
+'https://huggingface.co/lysandre/arxiv-nlp/resolve/877b84a8f93f2d619faa2a6e514a32beef88ab0a/config.json'
+```
+
+## `cached_download`
+
+The `cached_download` function is useful for downloading and caching a file on your local disk. Once it is stored in your cache, you don't have to redownload the file the next time you use it. This is a hands-free solution for staying up to date with new file versions. When one of your downloaded files is updated in the repository, it is automatically downloaded and stored for you.
+
+Begin by downloading your file with `hf_hub_url`, and then pass the specified URL to `cached_download`:
+
+```python
+>>> from huggingface_hub import hf_hub_url, cached_download
+>>> config_file_url = hf_hub_url("lysandre/arxiv-nlp", filename="config.json")
+>>> cached_download(config_file_url)
+'/home/lysandre/.cache/huggingface/hub/bc0e8cc2f8271b322304e8bb84b3b7580701d53a335ab2d75da19c249e2eeebb.066dae6fdb1e2b8cce60c35cc0f78ed1451d9b341c78de19f3ad469d10a8cbb1'
+```
+
+The `hf_hub_url` and `cached_download` functions work hand in hand to download a file. In fact, this is exactly how the `hf_hub_download` function from the tutorial works! `hf_hub_download` is simply a wrapper that calls both `hf_hub_url` and `cached_download`.
+
+## `snapshot_download`
+
+The `snapshot_download` function works well for downloading an entire repository. Like the previous functions, all downloaded files are cached on your local disk, and will be automatically updated if a file in the repository is changed.
```

Comment:
I think this method is a bit complex in the way it works - so I would advocate for adding a few additional details:
- The entire repository is fetched at a given revision (or `main`) and cached
- This means that a single file change will result in the entire repository needing to be redownloaded, and not just the file
- (Optional) This constraint means it's usually better to manually download all files using the methods above if the user is aware of which files they're going to need. This method is particularly useful when a library has its downloading utilities as a helper, unaware of which files need to be downloaded.